### PR TITLE
<regex>: Follow the Standard in regex_traits::transform()

### LINKED ISSUE #1004: <regex>: Error C2664 in std::regex_traits::transform
**Describe the bug**
`std::regex_traits::transform` tries to use `const _Elem*` as internal iterator instead of user-defined iterator.

**Command-line test case**
```
d:\Temp2>type repro.cpp
#include <regex>
#include <iostream>

template <class It>
class forward_iterator
{
    It it_;

    template <class U> friend class forward_iterator;
public:
    typedef          std::forward_iterator_tag                 iterator_category;
    typedef typename std::iterator_traits<It>::value_type      value_type;
    typedef typename std::iterator_traits<It>::difference_type difference_type;
    typedef It                                                 pointer;
    typedef typename std::iterator_traits<It>::reference       reference;

    It base() const { return it_; }

    forward_iterator() : it_() {}
    explicit forward_iterator(It it) : it_(it) {}
    template <class U>
    forward_iterator(const forward_iterator<U>& u) : it_(u.it_) {}

    reference operator*() const { return *it_; }
    pointer operator->() const { return it_; }

    forward_iterator& operator++() { ++it_; return *this; }
    forward_iterator operator++(int)
    {
        forward_iterator tmp(*this); ++(*this); return tmp;
    }

    friend bool operator==(const forward_iterator& x, const forward_iterator& y)
    {
        return x.it_ == y.it_;
    }
    friend bool operator!=(const forward_iterator& x, const forward_iterator& y)
    {
        return !(x == y);
    }

    template <class T>
    void operator,(T const &) = delete;
};

int main()
{
    std::regex_traits<char> t;
    const char a[] = "a";
    const char B[] = "B";
    typedef forward_iterator<const char*> F;
    bool r = t.transform(F(a), F(a + 1)) > t.transform(F(B), F(B + 1));
    (void) r;
}

d:\Temp2>cl /EHsc /W4 .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\regex(283): error C2664: 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::collate<_Elem>::transform(const _Elem *,const _Elem *) const': cannot convert argument 1 from '_FwdIt' to 'const _Elem *'
        with
        [
            _Elem=char
        ]
        and
        [
            _FwdIt=F
        ]
        and
        [
            _Elem=char
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\regex(283): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\locale(41): note: see declaration of 'std::collate<_Elem>::transform'
        with
        [
            _Elem=char
        ]
.\repro.cpp(52): note: see reference to function template instantiation 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::_Regex_traits<char>::transform<F>(_FwdIt,_FwdIt) const' being compiled
        with
        [
            _FwdIt=F
        ]
.\repro.cpp(52): note: see reference to function template instantiation 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::_Regex_traits<char>::transform<F>(_FwdIt,_FwdIt) const' being compiled
        with
        [
            _FwdIt=F
        ]
```

**Expected behavior**
The program should compile

**STL version**
```
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1
```

**Additional context**
The following is the equivalent of the repro:
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/tests/libcxx/expected_results.txt#L564-L565

Also tracked by DevCom-1076807 and Microsoft-internal VSO-1142061 / [AB#1142061](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1142061).

### PR DESCRIPTION
**Pull request by Arzaghi**
Fixes #1004

### CODE REVIEW

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this bug! I pushed two microscopic changes:

* The Standard depicts `str` but we need to use ugly `_Str`.
* Purely stylistic: although the Standard depicts `.length()` being called, we conventionally always call the synonymous `size()` for `basic_string`. I also changed one pre-existing occurrence nearby. Now, the only `length()` calls in the entire STL are for `sub_match`.

On my machine, the libcxx test is auto-skipped as unsupported due to lacking `LOCALE_cs_CZ_ISO8859_2`, but I manually verified that (without those lines) the test fails to compile before your change, and successfully compiles after your change. (We *might* need to revise `skipped_tests.txt` when we mirror this to the MSVC-internal repo, not sure, but I'll take care of that if necessary.)

File: stl/inc/regex [hunk 1]
```diff
@@ -280,7 +280,8 @@ public:
 
     template <class _FwdIt>
     string_type transform(_FwdIt _First, _FwdIt _Last) const { // apply locale-specific transformation
-        return _Getcoll()->transform(_First, _Last);
+        string_type _Str(_First, _Last);
```
Comment (CaseyCarter) [inline stl/inc/regex:283]: Peta-nitpick: In C++14-mode, `_Str.data()` returns a pointer to `const`, but in C++17-and-later a pointer-to-mutable. Formally, if someone links together a C++14 TU and a C++17 TU that instantiate the same specialization of this function template we're violating the One Definition Rule. I can't imagine this ODR violation causing actual problems, but it's easy enough to fix by making `_Str` a constant so the next line calls the same `data` in both C++14 and C++17 modes.
```suggestion
        const string_type _Str(_First, _Last);
```

Comment (CaseyCarter) [reply stl/inc/regex:283]: This is so ridiculously esoteric that I feel silly requesting changes, so I'm going to instead approve, apply the change myself, and hope nobody ever reads this ;)

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/regex (hunk 1 above)
Comment (CaseyCarter) [reply stl/inc/regex]: @StephanTLavavej, I should call your attention to the fact that I made this silly change after you approved.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this user-reported bug and removing another entry from our list of skipped libcxx tests! :beetle: :smile_cat:

**Pull request merged**